src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -4241,6 +4241,52 @@ FCVT_SAFE(fcvt_l_d, d);
 
 #undef FCVT_SAFE
 
+void MacroAssembler::java_round_float(Register dst, FloatRegister src,
+                                      FloatRegister ftmp, Register tmp) {
+  Label done;
+  assert_different_registers(dst, tmp);
+  mv(dst, zr);
+
+  fclass_s(tmp, src);
+
+  // dst = 0
+  // if +/-0, +/-subnormal numbers, signaling/quiet NaN
+  andi(tmp, tmp, 0b1100111100);
+  bnez(tmp, done);
+
+  // dst = (src + 0.5f) rounded down towards negative infinity
+  // if +/-inf, +/-normal numbers
+  li(tmp, 0x3f000000);
+  fmv_w_x(ftmp, tmp);
+  fadd_s(ftmp, src, ftmp);
+  fcvt_w_s(dst, ftmp, RoundingMode::rdn);
+
```

Comment:
This still doesn't look right to me. I urge you to test it against the Java implementation over the full 32-bit range.

Reply:
I think it may start working if rounding mode for fadd_s would be changed from default rne, to rdn

Reply:
It won't because of double rounding, and changing rounding modes is expensive. 
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L5882

Reply:
But doing fadd 0.5 to the number, which can't have fractional part, in rdn mode becomes no-op.
At least on single precision floats it works:
fadd(-8388609.0, +0.5, rdn) results in -8388609.0
and the mode for both fadd and fcvt will be the same, (perf tests showed no difference on thead tho)

Reply:
Maybe. I didn't try it, but on a great big out-of-order machine changing floating-point modes can be fantastically expensive, forcing ops in progress to retire, changing mode, and then continuing. Effectively it's as bad as a mispredict. Given that a correct solution that doesn't involve changing modes is available, I don't see why you wouldn't use it.

Reply:
> I have consulted with our h/w team and they told me next: multi-issue FP Unit can process few (data-independent) fp instructions at a time, even if they have different rounding mode. The only issue is when the rounding mode is set to dynamic rounding (aka get it from csr), but it's not our case here

OK. Do what you will, but you're coding for an architecture not an implementation, and this code may stand for many years.

Reply:
First I've tried to implement rounding on riscv similar to aarch64, separated negative and positive numbers but branching is too expensive and I got not much performance improvement against the current java implementation.

I like Vladimir's idea, the algorithm gives significant performance improvement and why not to take this advantage. But the above mentioned fix, related to fadd_s(), should be made. Not all numbers are processed correctly without it. Having changed rounding mode of fadd_s() we've got two sequential instructions with the same rounding mode  - rdn. 

The performance improvement with the last fix on the T-Head RVB-ICE board is as follow:
```
Benchmark                              (TESTSIZE)   Mode  Cnt    Score   Error   Units
FpRoundingBenchmark.test_round_double        2048  thrpt   15  111.278  0.349  ops/ms
FpRoundingBenchmark.test_round_float         2048  thrpt   15  115.776  0.323  ops/ms
```
@theRealAph As you suggested I've also tested this algorithm on the full 32-bit range [0; 0xFFFFFFFF] using Float.intBitsToFloat(x) and all the numbers were processed correctly. The output of this algorithm is equal to the current java Math.round() implementation output.

Reply:
```
In typical implementations, writes to the dynamic rounding mode CSR state will serialize
the pipeline.
Static rounding modes are used to implement specialized arithmetic operations that often
have to switch frequently between different rounding modes.
```
-- from `“F” Standard Extension`

Seems to me, the static rounding is fine in riscv arch?